includes/backup.js

```diff
@@ -13,267 +13,100 @@
-      ee.emit('received', {
-        batch: thisBatch,
-        data: output,
-        length: output.length,
-        time: t,
-        total: total
-      }, q, logCompletedBatch);
-    }).catch(err => {
-      if (!hasErrored) {
-        hasErrored = true;
-        err = error.convertResponseError(err);
-        // Kill the queue for fatal errors
-        q.kill();
-        ee.emit('error', err);
+        return logFileGetBatches(options.log, summary.batches);
+      } else {
+      // Not resuming, start from spooling changes
+        return [
+          ...spoolchanges(db, options.log, options.bufferSize),
+          new SideEffect((backupBatch) => {
+            ee.emit('changes', backupBatch.batch);
+          }) // Emit the user facing changes event for each batch of changes
+        ];
       }
-      done();
-    });
-  }, parallelism);
-
-  for (const i in batches) {
-    q.push(batches[i]);
-  }
-
-  q.drain(function() {
-    callback(null, { total: total });
-  });
-}
-
-/**
- * Returns first N properties on an object.
- *
- * @param {object} obj - object with properties
- * @param {number} count - number of properties to return
- */
-function getPropertyNames(obj, count) {
-  // decide which batch numbers to deal with
-  const batchestofetch = [];
-  let j = 0;
-  for (const i in obj) {
-    batchestofetch.push(parseInt(i));
-    j++;
-    if (j >= count) break;
-  }
-  return batchestofetch;
-}
+    })
+  // Create a pipeline of the source streams and the backup mappings
+    .then((srcStreams) => {
+      const backup = new Backup(db);
+      return pipeline(
+        ...srcStreams, // the source streams from the previous block (spool changes or resumed log)
+        new MappingStream(backup.pendingToFetched, options.parallelism), // fetch the batches at the configured concurrency
+        new WritableWithPassThrough(
+          'backup', // name for logging
+          targetStream, // backup file
+          null, // no need to write a last chunk
+          backup.backupBatchToBackupFileLine // map the backup batch to a string for the backup file
+        ), // WritableWithPassThrough writes the fetched docs to the backup file and passes on the result metadata
+        new DelegateWritable(
+          'logFileDoneWriter', // Name for debug
+          createWriteStream(options.log, { flags: 'a' }), // log file for appending
+          null, // no last chunk to write
+          backup.backupBatchToLogFileLine, // Map the backed up batch result to a log file "done" line
+          (backupBatch) => {
+            total += backupBatch.docs.length;
+            ee.emit('written', { total, time: (new Date().getTime() - start) / 1000, batch: backupBatch.batch });
+          } // post write function emits the written event
+        ) // DelegateWritable writes the log file done lines
+      );
+    })
+    .then(() => {
+      ee.emit('finished', { total });
+    })
+    .catch((e) => { throw convertResponseError(e); });
```

Comment:
That's a nit, but we don't need curly on one-liners or parens on one-args, so this can be written as `.catch(e => throw convertResponseError(e))`

Same with above, btw, are curlies there just to suppress return from `ee.emit` ?

Reply:
For the former: 93b4a2dd0acf7dc4d0e80cfb2a25eefdd01bde5b
For the latter:
`.catch(e => throw convertResponseError(e));`
gives me an
"Unexpected token throw"

Reply:
That's weird, that really should be a syntax only thing. Is it because event emitter now returns and that interpreted as promise reject and passed as `e`? I see `the boolean true was thrown, throw an Error :)` in CI and `emit` returns `true` when there are listeners

Reply:
ok dropping it, there is some follow on cleanup coming anyway